### PR TITLE
[KYUUBI #999] Build the basic framework for rest frontend service

### PR DESCRIPTION
**Pull request by yanghua**
### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 1]
```diff
@@ -379,6 +381,19 @@ object KyuubiConf {
     .transform(_.toLowerCase(Locale.ROOT))
     .createWithDefault(SaslQOP.AUTH.toString)
 
+  val FRONTEND_REST_BIND_HOST: OptionalConfigEntry[String] = buildConf("frontend.rest.bind.host")
+    .doc("Hostname or IP of the machine on which to run the REST frontend service.")
+    .version("1.0.0")
```
Comment (zwangsheng) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:386]: this version  may need to follow the version of Kyuubi

Comment (yanghua) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:386]: Hi, yes, this PR is under draft state. It's not ready for review. Opening for verifying CI.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 2]
```diff
@@ -132,6 +132,8 @@ case class KyuubiConf(loadSysDefault: Boolean = true) extends Logging {
   private val serverOnlyConfEntries: Set[ConfigEntry[_]] = Set(
     FRONTEND_BIND_HOST,
```
Comment (yanghua) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:133]: Do we have any strategy for deprecated config options?

Comment (zwangsheng) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:133]: Maybe you can refer to LEGACY_ENGINE_SHARE_LEVEL?

Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:133]: use `fallbackConf`

Comment (yanghua) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:133]: Got it

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/ConfigEntry.scala [hunk 3]
```diff
@@ -173,6 +173,29 @@ class ConfigEntryFallback[T](
   override def internal: Boolean = _internal
 }
 
+class OptionalConfigEntryFallback[T](
+                              _key: String,
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/ConfigEntry.scala:177]: The indentation looks weird here.

File: kyuubi-common/src/test/scala/org/apache/kyuubi/config/ConfigEntrySuite.scala [hunk 4]
```diff
@@ -142,4 +142,30 @@ class ConfigEntrySuite extends KyuubiFunSuite {
...
+    assert(fallback.version === "1.2.0")
+    assert(fallback.typ === "string")
+    assert(fallback.internal === false)
+    assert(fallback.toString === s"ConfigEntry(key=kyuubi.fallback.spark, " +
+      s"defaultValue=${ConfigEntry.UNDEFINED}, doc=fallback, version=1.2.0, type=string)")
+
+    val conf = KyuubiConf()
+    assert(conf.get(fallback) === None)
+    conf.set(origin.key, "new value")
+    assert(conf.get(fallback) === Option("new value"), "fallback to new original key")
```
Comment (pan3793) [inline kyuubi-common/src/test/scala/org/apache/kyuubi/config/ConfigEntrySuite.scala:169]: `Option("new value")` => `Some("new value")`

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/ConfigEntry.scala (hunk 3 above)
Comment (yanghua) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/ConfigEntry.scala:177]: Will fix it!

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 5]
```diff
@@ -442,6 +444,19 @@ object KyuubiConf {
...
+    .version("1.4.0")
+    .stringConf
+    .createOptional
+
+  val FRONTEND_REST_BIND_PORT: ConfigEntry[Int] = buildConf("frontend.rest.bind.port")
+    .doc("Port of the machine on which to run the REST frontend service.")
+    .version("1.4.0")
+    .intConf
+    .checkValue(p => p == 0 || (p > 1024 && p < 65535), "Invalid Port number")
+    .createWithDefault(10009)
```
Comment (yanghua) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:458]: Should we choose another default port to distinguish from the thrift binary bind port?

Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:458]: 10099?

Comment (yanghua) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:458]: sounds good.

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks merged to master

**Pull request closed without merging**